// @cubejs-backend/schema-compiler/dist/src/compiler/transpilers/CubePropContextTranspiler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CubePropContextTranspiler = void 0;
const t = __importStar(require("@babel/types"));
const ramda_1 = __importDefault(require("ramda"));
class CubePropContextTranspiler {
    constructor(cubeSymbols, cubeDictionary) {
        this.cubeSymbols = cubeSymbols;
        this.cubeDictionary = cubeDictionary;
    }
    traverseObject() {
        const self = this;
        return {
            CallExpression(path) {
                const args = path.get('arguments');
                if (path.node.callee && path.node.callee.type === 'Identifier' && (path.node.callee.name === 'view' || path.node.callee.name === 'cube')) {
                    if (args && args[args.length - 1]) {
                        const cubeName = args[0].node.type === 'StringLiteral' && args[0].node.value ||
                            args[0].node.type === 'TemplateLiteral' &&
                                args[0].node.quasis.length &&
                                args[0].node.quasis[0].value.cooked;
                        args[args.length - 1].traverse(self.sqlAndReferencesFieldVisitor(cubeName));
                        args[args.length - 1].traverse(self.knownIdentifiersInjectVisitor('extends', name => self.cubeDictionary.resolveCube(name)));
                    }
                    // @ts-ignore @todo Unsafely?
                }
                else if (path.node.callee.name === 'context') {
                    args[args.length - 1].traverse(self.sqlAndReferencesFieldVisitor(null));
                    // @ts-ignore @todo Unsafely?
                }
                else if (path.node.callee.name === 'dashboardTemplate') {
                    args[args.length - 1].traverse(self.shortNamedReferencesFieldVisitor(null));
                }
            }
        };
    }
    sqlAndReferencesFieldVisitor(cubeName) {
        return this.knownIdentifiersInjectVisitor(/^(sql|measureReferences|dimensionReferences|segmentReferences|timeDimensionReference|rollupReferences|drillMembers|drillMemberReferences|contextMembers|columns)$/, name => this.cubeSymbols.resolveSymbol(cubeName, name) || this.cubeSymbols.isCurrentCube(name));
    }
    shortNamedReferencesFieldVisitor(cubeName) {
        return this.knownIdentifiersInjectVisitor(/^.*(measures|dimensions|segments|measure|dimension|segment|member)$/, name => this.cubeSymbols.resolveSymbol(cubeName, name) || this.cubeSymbols.isCurrentCube(name));
    }
    knownIdentifiersInjectVisitor(field, resolveSymbol) {
        const self = this;
        return {
            ObjectProperty(path) {
                if (path.node.key.type === 'Identifier' && path.node.key.name.match(field)) {
                    const knownIds = self.collectKnownIdentifiers(resolveSymbol, path.get('value'));
                    path.get('value').replaceWith(t.arrowFunctionExpression(knownIds.map(i => t.identifier(i)), path.node.value, false));
                }
            }
        };
    }
    collectKnownIdentifiers(resolveSymbol, path) {
        const identifiers = [];
        const self = this;
        if (path.node.type === 'Identifier') {
            this.matchAndPushIdentifier(path, resolveSymbol, identifiers);
        }
        path.traverse({
            Identifier(p) {
                self.matchAndPushIdentifier(p, resolveSymbol, identifiers);
            }
        });
        return ramda_1.default.uniq(identifiers);
    }
    matchAndPushIdentifier(path, resolveSymbol, identifiers) {
        if ((!path.parent ||
            (path.parent.type !== 'MemberExpression' || path.parent.type === 'MemberExpression' && path.key !== 'property')) &&
            resolveSymbol(path.node.name)) {
            identifiers.push(path.node.name);
        }
    }
}
exports.CubePropContextTranspiler = CubePropContextTranspiler;
