// @cubejs-backend/server-core/dist/src/core/DevServer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevServer = void 0;
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const api_gateway_1 = require("@cubejs-backend/api-gateway");
const cloud_1 = require("@cubejs-backend/cloud");
const templates_1 = require("@cubejs-backend/templates");
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const is_docker_1 = __importDefault(require("is-docker"));
const server_1 = require("./server");
const repo = {
    owner: 'cube-js',
    name: 'cubejs-playground-templates'
};
class DevServer {
    constructor(cubejsServer) {
        this.cubejsServer = cubejsServer;
        this.applyTemplatePackagesPromise = null;
        this.dashboardAppProcess = null;
        this.livePreviewWatcher = new cloud_1.LivePreviewWatcher();
    }
    initDevEnv(app, options) {
        const port = process.env.PORT || 4000; // TODO
        const apiUrl = process.env.CUBEJS_API_URL || `http://localhost:${port}`;
        // todo: empty/default `apiSecret` in dev mode to allow the DB connection wizard
        const cubejsToken = jsonwebtoken_1.default.sign({}, options.apiSecret || 'secret', { expiresIn: '1d' });
        if (process.env.NODE_ENV !== 'production') {
            console.log('🔓 Authentication checks are disabled in developer mode. Please use NODE_ENV=production to enable it.');
        }
        else {
            console.log(`🔒 Your temporary cube.js token: ${cubejsToken}`);
        }
        console.log(`🦅 Dev environment available at ${apiUrl}`);
        this.cubejsServer.event('Dev Server Start');
        const serveStatic = require('serve-static');
        const catchErrors = (handler) => async (req, res, next) => {
            try {
                await handler(req, res, next);
            }
            catch (e) {
                console.error((e.stack || e).toString());
                this.cubejsServer.event('Dev Server Error', { error: (e.stack || e).toString() });
                res.status(500).json({ error: (e.stack || e).toString() });
            }
        };
        app.get('/playground/context', catchErrors((req, res) => {
            this.cubejsServer.event('Dev Server Env Open');
            res.json({
                cubejsToken,
                apiUrl: process.env.CUBEJS_API_URL,
                basePath: options.basePath,
                anonymousId: this.cubejsServer.anonymousId,
                coreServerVersion: this.cubejsServer.coreServerVersion,
                projectFingerprint: this.cubejsServer.projectFingerprint,
                shouldStartConnectionWizardFlow: !this.cubejsServer.configFileExists(),
                livePreview: options.livePreview
            });
        }));
        app.get('/playground/db-schema', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server DB Schema Load');
            const driver = await this.cubejsServer.getDriver({
                dataSource: req.body.dataSource || 'default',
                authInfo: null,
                securityContext: null,
                requestId: api_gateway_1.getRequestIdFromRequest(req),
            });
            const tablesSchema = await driver.tablesSchema();
            this.cubejsServer.event('Dev Server DB Schema Load Success');
            if (Object.keys(tablesSchema || {}).length === 0) {
                this.cubejsServer.event('Dev Server DB Schema Load Empty');
            }
            res.json({ tablesSchema });
        }));
        app.get('/playground/files', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server Files Load');
            const files = await this.cubejsServer.repository.dataSchemaFiles();
            res.json({
                files: files.map(f => ({
                    ...f,
                    absPath: path_1.default.resolve(path_1.default.join(this.cubejsServer.repository.localPath(), f.fileName))
                }))
            });
        }));
        app.post('/playground/generate-schema', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server Generate Schema');
            if (!req.body) {
                throw new Error('Your express app config is missing body-parser middleware. Typical config can look like: `app.use(bodyParser.json({ limit: \'50mb\' }));`');
            }
            if (!req.body.tables) {
                throw new Error('You have to select at least one table');
            }
            const dataSource = req.body.dataSource || 'default';
            const driver = await this.cubejsServer.getDriver({
                dataSource,
                authInfo: null,
                securityContext: null,
                requestId: api_gateway_1.getRequestIdFromRequest(req),
            });
            const tablesSchema = req.body.tablesSchema || (await driver.tablesSchema());
            const ScaffoldingTemplate = require('@cubejs-backend/schema-compiler/scaffolding/ScaffoldingTemplate');
            const scaffoldingTemplate = new ScaffoldingTemplate(tablesSchema, driver);
            const files = scaffoldingTemplate.generateFilesByTableNames(req.body.tables, { dataSource });
            const schemaPath = options.schemaPath || 'schema';
            await Promise.all(files.map(file => fs_extra_1.default.writeFile(path_1.default.join(schemaPath, file.fileName), file.content)));
            res.json({ files });
        }));
        let lastApplyTemplatePackagesError = null;
        app.get('/playground/dashboard-app-create-status', catchErrors(async (req, res) => {
            const sourcePath = path_1.default.join(options.dashboardAppPath, 'src');
            if (lastApplyTemplatePackagesError) {
                const toThrow = lastApplyTemplatePackagesError;
                lastApplyTemplatePackagesError = null;
                throw toThrow;
            }
            if (this.applyTemplatePackagesPromise) {
                if (req.query.instant) {
                    res.status(404).json({ error: 'Dashboard app creating' });
                    return;
                }
                await this.applyTemplatePackagesPromise;
            }
            // docker-compose share a volume for /dashboard-app and directory will be empty
            if (!fs_extra_1.default.pathExistsSync(options.dashboardAppPath) || fs_extra_1.default.readdirSync(options.dashboardAppPath).length === 0) {
                res.status(404).json({
                    error: `Dashboard app not found in '${path_1.default.resolve(options.dashboardAppPath)}' directory`
                });
                return;
            }
            if (!fs_extra_1.default.pathExistsSync(sourcePath)) {
                res.status(404).json({
                    error: `Dashboard app corrupted. Please remove '${path_1.default.resolve(options.dashboardAppPath)}' directory and recreate it`
                });
                return;
            }
            res.json({
                status: 'created',
                installedTemplates: templates_1.AppContainer.getPackageVersions(options.dashboardAppPath)
            });
        }));
        app.get('/playground/start-dashboard-app', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server Start Dashboard App');
            if (!this.dashboardAppProcess) {
                const { dashboardAppPort = 3000 } = options;
                this.dashboardAppProcess = cross_spawn_1.default('npm', [
                    'run',
                    'start',
                    '--',
                    '--port',
                    dashboardAppPort.toString(),
                    ...(is_docker_1.default() ? ['--host', '0.0.0.0'] : [])
                ], {
                    cwd: options.dashboardAppPath,
                    env: {
                        ...process.env,
                        PORT: dashboardAppPort
                    }
                });
                this.dashboardAppProcess.dashboardUrlPromise = new Promise((resolve) => {
                    this.dashboardAppProcess.stdout.on('data', (data) => {
                        console.log(data.toString());
                        if (data.toString().match(/Compiled/)) {
                            resolve(options.dashboardAppPort);
                        }
                    });
                });
                this.dashboardAppProcess.on('close', exitCode => {
                    if (exitCode !== 0) {
                        console.log(`Dashboard react-app failed with exit code ${exitCode}`);
                        this.cubejsServer.event('Dev Server Dashboard App Failed', { exitCode });
                    }
                    this.dashboardAppProcess = null;
                });
            }
            await this.dashboardAppProcess.dashboardUrlPromise;
            res.json({ dashboardPort: options.dashboardAppPort });
        }));
        app.get('/playground/dashboard-app-status', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server Dashboard App Status');
            const dashboardPort = this.dashboardAppProcess && await this.dashboardAppProcess.dashboardUrlPromise;
            res.json({
                running: !!dashboardPort,
                dashboardPort,
                dashboardAppPath: path_1.default.resolve(options.dashboardAppPath)
            });
        }));
        app.post('/playground/apply-template-packages', catchErrors(async (req, res) => {
            this.cubejsServer.event('Dev Server Download Template Packages');
            const fetcher = process.env.TEST_TEMPLATES ? new templates_1.DevPackageFetcher(repo) : new templates_1.PackageFetcher(repo);
            this.cubejsServer.event('Dev Server App File Write');
            const { toApply, templateConfig } = req.body;
            const applyTemplates = async () => {
                const manifestJson = await fetcher.manifestJSON();
                const response = await fetcher.downloadPackages();
                let templatePackages = [];
                if (typeof toApply === 'string') {
                    const template = manifestJson.templates.find(({ name }) => name === toApply);
                    templatePackages = template.templatePackages;
                }
                else {
                    templatePackages = toApply;
                }
                const dt = new templates_1.DependencyTree(manifestJson, templatePackages);
                const appContainer = new templates_1.AppContainer(dt.getRootNode(), {
                    appPath: options.dashboardAppPath,
                    packagesPath: response.packagesPath
                }, templateConfig);
                this.cubejsServer.event('Dev Server Create Dashboard App');
                await appContainer.applyTemplates();
                this.cubejsServer.event('Dev Server Create Dashboard App Success');
                this.cubejsServer.event('Dev Server Dashboard Npm Install');
                await appContainer.ensureDependencies();
                this.cubejsServer.event('Dev Server Dashboard Npm Install Success');
                // todo: uncomment
                // fetcher.cleanup();
            };
            if (this.applyTemplatePackagesPromise) {
                this.applyTemplatePackagesPromise = this.applyTemplatePackagesPromise.then(applyTemplates);
            }
            else {
                this.applyTemplatePackagesPromise = applyTemplates();
            }
            const promise = this.applyTemplatePackagesPromise;
            promise.then(() => {
                if (promise === this.applyTemplatePackagesPromise) {
                    this.applyTemplatePackagesPromise = null;
                }
            }, (err) => {
                console.log('err', err);
                lastApplyTemplatePackagesError = err;
                if (promise === this.applyTemplatePackagesPromise) {
                    this.applyTemplatePackagesPromise = null;
                }
            });
            res.json(true); // TODO
        }));
        app.get('/playground/manifest', catchErrors(async (_, res) => {
            const fetcher = process.env.TEST_TEMPLATES ? new templates_1.DevPackageFetcher(repo) : new templates_1.PackageFetcher(repo);
            res.json(await fetcher.manifestJSON());
        }));
        app.get('/playground/live-preview/start/:token', catchErrors(async (req, res) => {
            this.livePreviewWatcher.setAuth(req.params.token);
            this.livePreviewWatcher.startWatch();
            res.setHeader('Content-Type', 'text/html');
            res.write('<html><head><script>window.close();</script></body></html>');
            res.end();
        }));
        app.get('/playground/live-preview/stop', catchErrors(async (req, res) => {
            this.livePreviewWatcher.stopWatch();
            res.json({ active: false });
        }));
        app.get('/playground/live-preview/status', catchErrors(async (req, res) => {
            const statusObj = await this.livePreviewWatcher.getStatus();
            res.json(statusObj);
        }));
        app.post('/playground/live-preview/token', catchErrors(async (req, res) => {
            const token = await this.livePreviewWatcher.createTokenWithPayload(req.body);
            res.json({ token });
        }));
        app.use(serveStatic(path_1.default.join(__dirname, '../../../playground'), {
            lastModified: false,
            etag: false,
            setHeaders: (res, url) => {
                if (url.indexOf('/index.html') !== -1) {
                    res.setHeader('Cache-Control', 'no-cache');
                }
            }
        }));
        app.post('/playground/test-connection', catchErrors(async (req, res) => {
            const { variables = {} } = req.body || {};
            let driver = null;
            try {
                if (!variables.CUBEJS_DB_TYPE) {
                    throw new Error('CUBEJS_DB_TYPE is required');
                }
                // Backup env variables for restoring
                const originalProcessEnv = process.env;
                process.env = {
                    ...process.env,
                };
                for (const [envName, value] of Object.entries(variables)) {
                    process.env[envName] = value;
                }
                driver = server_1.CubejsServerCore.createDriver(variables.CUBEJS_DB_TYPE);
                // Restore original process.env
                process.env = originalProcessEnv;
                await driver.testConnection();
                this.cubejsServer.event('test_database_connection_success');
                return res.json('ok');
            }
            catch (error) {
                this.cubejsServer.event('test_database_connection_error');
                return res.status(400).json({
                    error: error.toString()
                });
            }
            finally {
                if (driver && driver.release) {
                    await driver.release();
                }
            }
        }));
        app.get('/restart', catchErrors(async (_, res) => {
            process.kill(process.pid, 'SIGUSR1');
            return res.json('Restarting...');
        }));
        app.post('/playground/env', catchErrors(async (req, res) => {
            let { variables = {} } = req.body || {};
            if (!variables.CUBEJS_API_SECRET) {
                variables.CUBEJS_API_SECRET = options.apiSecret;
            }
            variables = Object.entries(variables).map(([key, value]) => ([key, value].join('=')));
            if (fs_extra_1.default.existsSync('./.env')) {
                fs_extra_1.default.removeSync('./.env');
            }
            if (!fs_extra_1.default.existsSync('./schema')) {
                fs_extra_1.default.mkdirSync('./schema');
            }
            fs_extra_1.default.writeFileSync('.env', variables.join('\n'));
            res.status(200).json('ok');
        }));
        app.post('/playground/token', catchErrors(async (req, res) => {
            const { payload = {} } = req.body;
            const jwtOptions = typeof payload.exp != null ? {} : { expiresIn: '1d' };
            const token = jsonwebtoken_1.default.sign(payload, options.apiSecret, jwtOptions);
            res.json({ token });
        }));
    }
}
exports.DevServer = DevServer;
